core/Pos/Binary/Class/Primitive.hs

```diff
@@ -3,110 +3,135 @@
-       , getBytes
-       , putBytes
-
-       , getAsciiString1b
-       , putAsciiString1b
-       , sizeAsciiString1b
-
-       , Raw (..)
-       ) where
-
+    ( serialize
+    , serialize'
+    , deserialize
+    , deserialize'
+    , deserializeOrFail
+    , deserializeOrFail'
+    , putCopyBi
+    , getCopyBi
+    , Raw(..)
+    -- * Binary serialization
+    , AsBinary (..)
+    , AsBinaryClass (..)
+    , fromBinaryM
+    -- * Temporary functions
+    , biSize
+    -- * Backward-compatible functions
+    , decodeFull
+    ) where
+
+import qualified Codec.CBOR.Read                  as CBOR.Read
+import qualified Codec.CBOR.Write                 as CBOR.Write
+import           Control.Exception                (throw)
+import           Control.Monad.ST                 (ST, runST)
+import qualified Data.ByteString                  as BS
+import qualified Data.ByteString.Lazy             as BSL
+import qualified Data.ByteString.Lazy.Internal    as BSL
+import           Data.SafeCopy                    (Contained, SafeCopy (..), contain, safeGet,
+                                                   safePut)
+import qualified Data.Serialize                   as Cereal (Get, Put)
+import           Data.Typeable                    (typeRep)
+
+import           Pos.Binary.Class.Core            (Bi(..))
 import           Universum
+import           Serokell.Data.Memory.Units       (Byte)
 
-import qualified Data.ByteString            as BS
-import           Data.Char                  (isAscii)
-import           Data.Reflection            (reifyNat)
-import           Data.SafeCopy              (Contained, SafeCopy (..), contain, safeGet,
-                                             safePut)
-import qualified Data.Serialize             as Cereal (Get, Put)
-import qualified Data.Store.Core            as Store
-import qualified Data.Store.Internal        as Store
-import           Formatting                 (formatToString, int, (%))
-
-import           Pos.Binary.Class.Core      (Bi (..), decodeFull, encode, isolate64Full,
-                                             limitGet)
-import           Pos.Binary.Class.Numbers   (TinyVarInt (..), UnsignedVarInt (..),
-                                             getWord8, putWord8)
-import           Pos.Binary.Class.Store     (Peek, Poke, PokeWithSize (..), Size (..),
-                                             StaticSize (..), putS)
-
-import           Pos.Binary.Class.Instances ()
--- TODO ^ Move Raw in Pos.Util.Util and remove the import.
-
--- | See 'putWord8'.
-putWord8S :: Word8 -> PokeWithSize ()
-putWord8S = putS @Word8
-
--- | See 'putBytes'.
-putBytesS :: ByteString -> PokeWithSize ()
-putBytesS bs = PokeWithSize (BS.length bs) (putBytes bs)
+-- | Serialize a Haskell value to an external binary representation.
+--
+-- The output is represented as a lazy 'BSL.ByteString' and is constructed
+-- incrementally.
+serialize :: Bi a => a -> BSL.ByteString
```

Comment:
The reviewer might get a bit confused by naming at first, but essentially now `encode` has a different type from what it used to have before, as it's a method of the `Bi` typeclass itself and it has type `encode :: a -> Encoding`. The migration cheatsheet is the following:

Old  | New
-----|-----
encode | serialize'
decodeFull | decodeFull
decodeOrFail | deserialize'